### ISSUE TITLE
Gnome 40 support

### ISSUE THREAD

**Issue by leonardehrenfried** [gTile/gTile#205]
Gnome 40 has been released.

I would love to have support for it.

Thanks!

**Comment by scherepanov** [comment]
You are not the only one!

**Comment by leonardehrenfried** [comment]
I took the hint ;)and have started looking into it.

If you add 40 to the list of compatible shell versions you can run the extension.

![Screenshot from 2021-03-29 16-08-22](https://user-images.githubusercontent.com/151346/112849249-289a3680-90a9-11eb-8201-588d811f849a.png)

However, nothing happens when you click on the top bar icon but also no errors are visible in the log.

I did see the following though:

```
gTile-Message: 16:06:24.452: WARNING: addSignalMethods is replacing existing [object Object] connect method
gTile-Message: 16:06:24.452: WARNING: addSignalMethods is replacing existing [object Object] disconnect method
gTile-Message: 16:06:24.452: WARNING: addSignalMethods is replacing existing [object Object] emit method
gTile-Message: 16:06:24.453: WARNING: addSignalMethods is replacing existing [object Object] connect method
gTile-Message: 16:06:24.453: WARNING: addSignalMethods is replacing existing [object Object] disconnect method
gTile-Message: 16:06:24.453: WARNING: addSignalMethods is replacing existing [object Object] emit method
gTile-Message: 16:06:24.454: WARNING: addSignalMethods is replacing existing [object Object] connect method
gTile-Message: 16:06:24.454: WARNING: addSignalMethods is replacing existing [object Object] disconnect method
gTile-Message: 16:06:24.454: WARNING: addSignalMethods is replacing existing [object Object] emit method
```
Is this problematic or to be expected?

**Comment by scherepanov** [comment]
Generally, speaking, it is a problem if any functionality is missing.
And, from your observations, gTile is not working at all, I would consider it a problem :-)
Log entries by themselves are not a problem. But, usually log entries are indication of some problem. It is a goal not to have any log entries from gTile.

**Comment by leonardehrenfried** [comment]
Actually, after doing a bit more testing, it _does_ work in Gnome 40!

![Screencast from 29-03-21 16 26 52](https://user-images.githubusercontent.com/151346/112852222-048c2480-90ac-11eb-8f18-8d44d5a3db7d.gif)

The only that doesn't is the prefs dialog but that is to be expected: https://gjs.guide/extensions/upgrading/gnome-shell-40.html#prefs

Would you like me to create a PR adding the version to `metadata.json`?

**Comment by scherepanov** [comment]
There is no reason to add v 40 to supported version, unless gTile is completely operational under v 40.
gTile is only partially working under v 40, you just found that prefs dialog is not there at all.
Users, who want to run incomplete gTile under v40, can do it by themselves.
Short answer - no! I will not accept PR with only added version to metadata.json.
It will be a complete fix for v 40, just will take loner.

**Comment by ijuma** [comment]
> Would you like me to create a PR adding the version to `metadata.json`?

Thanks for pointing this out. I am using gTile with Gnome 40 now and it's very helpful (even if the preference dialog doesn't work yet).

**Comment by cryptoquick** [comment]
Isn't getting something out sooner that works better than something that doesn't work at all?

In the meantime, assume I don't have technical context, how do I make this edit?

**Comment by cryptoquick** [comment]
For the meantime, I forked this project with partial 40 version support. Just check out:
https://github.com/cryptoquick/gTile40
Install bazel for your distro, as the README instructs.
Then run:
`bazel run :install-extension`
Reboot, and things should be back to working again.
Hope this helps others.

**Comment by StephanDollberg** [comment]
Any update on this? 

Fedora 34 now ships with Gnome 40 as well so this will be important for a lot more people now.

**Comment by scherepanov** [comment]
Generally, you can install gTile from master, and bump up version no to 40 in metadata.json.
Preferences will be failing, but rest is functional.
I do not have spare PC for Fedora beta. Will look on fixing preferences and doing proper release, when Fedora 34 will be officially released.
It is quite possible that Fedora 34 release will allow extensions to run without listing 40 as supported version of gnome-shell. At least, this is what happens in past. Then it will be a transparent upgrade (well, preferences will have to be fixed).
Patience!
Alternatively, find bug and submit PR.

**Comment by ijuma** [comment]
Fedora 34 was released today. :) https://fedoramagazine.org/announcing-fedora-34/

**Comment by scherepanov** [comment]
That is true! Fedora 34 with gnome-shell 40 is officially out as today morning.
I will take part-time vacation next week, will look on fixing preferences for gnome-shell v 40.

**Comment by StephanDollberg** [comment]
@scherepanov I kind of accepted your challenge and made it work (minus some optical details).

I hope it will help you, you can find it at https://github.com/gTile/gTile/compare/master...StephanDollberg:stephandollberg/gnome-40?expand=1

The only thing which I haven't found a replacement for is the margin stuff which is only commented out for now:

![image](https://user-images.githubusercontent.com/2904723/116320861-f0256f80-a7b0-11eb-942a-f22fa4afac4c.png)

I mostly used https://developer.gnome.org/gtk4/4.0/ch41s02.html#id-1.7.4.4.24

**Comment by scherepanov** [comment]
Great news, and big thanks for help.
If you will disable any functionality, you will always upset some users.
Exactly now #200 when users (rightfully) expect that missing functionality on margins should be restored.
In widely used GUI element like gTile, we have to keep all functionality in place. 
I will test and merge your changes. 
But, to roll out as version, we need also fix margins issue, including #200. Fortunately, fix for it already there (see this ticket), need to take from other branch, test and merge into master.
If you have time and can do it, would be of great speedup.

**Comment by StephanDollberg** [comment]
@scherepanov So the margin thing I was referring is only about margins in the preferences dialog if I am not missing something? You can see in my screenshot that there is margin at the borders of the window.

From what I can tell from #200 that's about margin when actually applying presets to your windows.

**Comment by StephanDollberg** [comment]
Have pushed some further changes to fix the margins in the prefs pane. Looks ok to me now.

**Comment by scherepanov** [comment]
That is perfect! Big thanks!
To make it proper way, can you submit PR?
It is a couple clicks on your branch.

**Comment by StephanDollberg** [comment]
Yep, will do shortly. Just noticed an issue where all keys in the accel tab show as "Invalid" even though they are actually working.

Suspect it's something with the key parser that changed in gtk4.

**Comment by StephanDollberg** [comment]
Ok, that is sorted as well. Happy to open a PR now.

The big question in the room is though how you'd want to merge that? 

I haven't really tested but I am sure some of those changes are not compatible with gtk3 anymore. The last change definitely isn't. So anything before gnome 40 won't really work.

I guess you could say that for older versions of gnome one should use an older version? I am not sure what you prefer.

There would also be the possibility of using something like get_major_version() to differentiate at places but not sure how much work that would be. Would need to double check differences for that (and also get a VM to test).

**Comment by scherepanov** [comment]
That is a bad news.
Vast majority of users are on pre-40 version on gnome-shell. Rolling out new version for them will be a breaking change.
Only option is to list new version as compatible with gnome-shell v 40. 
It will work, though I do not like it. Users on older versions of gnome-shell will be cut from fixes and new features etc.

Continuous breaking changing in gnome-shell are rather frustrating.

**Comment by StephanDollberg** [comment]
Ok, I will have a look. I hope it should be doable to support both at the same time. 

Give me a bit.

**Comment by StephanDollberg** [comment]
Have pushed a version now which should support both. Just downloading fedora 33 to test on older gnome

**Comment by StephanDollberg** [comment]
Opened PR https://github.com/gTile/gTile/pull/207

Working on both now.

**Comment by scherepanov** [comment]
Thanks!
Will try to test late tonight on old version, upgrade tomorrow and test on new version.
This is the plan.

**Comment by scherepanov** [comment]
Fixed in pr #207

**Comment by scherepanov** [comment]
New release of gTile with gnome-shell v 40.0 support is pending review on EGO
